checks.d/sidekiq.py

```diff
@@ -0,0 +1,246 @@
+    #   queue length 1,
+    #   queue length 2,
+    #   ...,
+    #   queue length n,
+    #   queue latency 1,
+    #   queue latency 2,
+    #   ...,
+    #   queue latency n
+    # ]
+    #
+    def _register_script(self):
+        s = """
+        local processed = redis.call("GET", KEYS[1])
+        local failed = redis.call("GET", KEYS[2])
+        local scheduled_size = redis.call("ZCARD", KEYS[3])
+        local retry_size = redis.call("ZCARD", KEYS[4])
+        local dead_size = redis.call("ZCARD", KEYS[5])
+        local processes_size = redis.call("SCARD", KEYS[6])
+        local processes = redis.call("SMEMBERS", KEYS[6])
+
+        local workers_size = 0
+        for i, pkey in ipairs(processes) do
+            local nspkey = ARGV[1] == "None" and pkey or (ARGV[1] .. ":" .. pkey)
+            workers_size = workers_size + tonumber(redis.call("HGET", nspkey, "busy"))
+        end
+
+        local queues = redis.call("SMEMBERS", KEYS[7])
+        local queue_lengths = {}
+        local queue_latencies = {}
+        for i, qkey in ipairs(queues) do
+            local nsqkey = "queue:" .. qkey
+            nsqkey = ARGV[1] == "None" and nsqkey or (ARGV[1] .. ":" .. nsqkey)
+
+            queue_lengths[qkey] = redis.call("LLEN", nsqkey)
+            local last = redis.call("LRANGE", nsqkey, -1, -1)[0]
+            if not last == nil then
+                queue_latencies[qkey] = tonumber(cjson.decode(last)["enqueued_at"])
+            else
+                queue_latencies[qkey] = 0.0
+            end
+        end
+
+        local offset = 8
+        local result = {processed,failed,scheduled_size,retry_size,dead_size,processes_size,workers_size}
+        local n = table.getn(queues)
+        result[offset] = n
+        for i, q in ipairs(queues) do
+            result[offset + i] = q
+            result[offset + i + n] = queue_lengths[q]
+            result[offset + i + n + n] = queue_latencies[q]
+        end
+        return result
+        """
+        return self.conn.register_script(s)
+
+    def _run_script(self):
+        return self._script(keys=[
+                            self.key('stat:processed'),
+                            self.key('stat:failed'),
+                            self.key('schedule'),
+                            self.key('retry'),
+                            self.key('dead'),
+                            self.key('processes'),
+                            self.key('queues')],
+                            args=[
+                                self.namespace])
+
+    def _check_stats(self):
+        res = self._run_script()
+        self.processed = int(res[0])
+        self.failed = int(res[1])
+        self.scheduled_size = res[2]
+        self.retry_size = res[3]
+        self.dead_size = res[4]
+        self.processes_size = res[5]
+        self.workers_size = res[6]
+        num_queues = res[7]
+
+        offset = 8
+        self.queue_stats = {}
+        for i in range(num_queues):
```

Comment:
Also looks like you could use generator between here and when you actually send the metrics. This way you wouldn't have to actually store all of these data.